File: etrike_protocol/codecs/typescript/generated.ts
import DISCOVERY from "../../generated/discovery.json";
import { type CanFrame, type CodecStatus, type DecodeResult, type EncodeResult, frame, validateFrame } from "./types";

interface FieldMetadata {
  readonly key: string;
  readonly byte: number;
  readonly bit: number;
  readonly bits: number;
  readonly signed?: boolean;
  readonly min?: number;
  readonly max?: number;
  readonly factor?: number;
  readonly offset?: number;
  readonly constant?: number;
  readonly enum?: Readonly<Record<string, string>>;
}

interface InstanceMetadata {
  readonly bus: string;
  readonly id: number;
  readonly frame_format: "standard" | "extended";
}

interface MessageMetadata {
  readonly byte_order: "big" | "little";
  readonly codec: { readonly strategy: "generated" | "custom" };
  readonly dlc: number;
  readonly instances: readonly InstanceMetadata[];
  readonly layout: { readonly kind: string; readonly fields?: readonly FieldMetadata[] };
}

interface DiscoveryMessage extends Omit<MessageMetadata, "instances"> {
  readonly canonical_key: string;
  readonly instances: readonly (Omit<InstanceMetadata, "id"> & { readonly id: string })[];
}

const metadata: Readonly<Record<string, MessageMetadata>> = Object.fromEntries(
  (DISCOVERY.messages as unknown as readonly DiscoveryMessage[]).map((message) => [
    message.canonical_key,
    {
      ...message,
      instances: message.instances.map((instance) => ({ ...instance, id: Number(instance.id) })),
    },
  ]),
);

export function hasMessage(message: string): boolean {
  return metadata[message] !== undefined;
}

export function isGenerated(message: string): boolean {
  return metadata[message]?.codec.strategy === "generated";
}

function generatedMessage(message: string): MessageMetadata {
  const selected = metadata[message];
  if (selected === undefined) throw new RangeError(`unknown message: ${message}`);
  if (selected.codec.strategy !== "generated") throw new RangeError(`${message} selects a custom codec`);
  return selected;
}

function instanceFor(message: MessageMetadata, bus: string): InstanceMetadata | undefined {
  const matches = message.instances.filter((instance) => instance.bus === bus);
  return matches.length === 1 ? matches[0] : undefined;
}

function limits(field: FieldMetadata): readonly [number, number] {
  const rawMinimum = field.signed ? -(2 ** (field.bits - 1)) : 0;
  const rawMaximum = field.signed ? 2 ** (field.bits - 1) - 1 : 2 ** field.bits - 1;
  const factor = field.factor ?? 1;
  const offset = field.offset ?? 0;
  return [field.min ?? rawMinimum * factor + offset, field.max ?? rawMaximum * factor + offset];
}

function isExhaustiveEnum(field: FieldMetadata): boolean {
  if (field.enum === undefined) return false;
  const rawMinimum = field.signed ? -(2 ** (field.bits - 1)) : 0;
  const rawMaximum = field.signed ? 2 ** (field.bits - 1) - 1 : 2 ** field.bits - 1;
  const factor = field.factor ?? 1;
  const offset = field.offset ?? 0;
  const rawMinVal = field.min !== undefined ? Math.round((field.min - offset) / factor) : rawMinimum;
  const rawMaxVal = field.max !== undefined ? Math.round((field.max - offset) / factor) : rawMaximum;
  const range = rawMaxVal - rawMinVal + 1;
  return Object.keys(field.enum).length >= range;
}

function encodePayload(message: MessageMetadata, values: Readonly<Record<string, unknown>>): DecodeResult<Uint8Array> {
  const payload = new Uint8Array(message.dlc);
  for (const field of message.layout.fields ?? []) {
    const value = field.constant ?? values[field.key];
    if (typeof value !== "number" || !Number.isFinite(value)) return ["value_out_of_range"];
    const [minimum, maximum] = limits(field);
    if (value < minimum || value > maximum) return ["value_out_of_range"];
    const unrounded = (value - (field.offset ?? 0)) / (field.factor ?? 1);
    let raw = Math.round(unrounded);
    if (Math.abs(unrounded - raw) > 1e-9) return ["value_out_of_range"];
    if (field.enum !== undefined && isExhaustiveEnum(field) && !(String(raw) in field.enum)) return ["invalid_enum"];
    if (field.signed && raw < 0) raw += 2 ** field.bits;

    if (field.bit === 0 && field.bits % 8 === 0) {
      const width = field.bits / 8;
      for (let index = 0; index < width; index += 1) {
        const shift = message.byte_order === "big" ? width - index - 1 : index;
        payload[field.byte + index] = Math.floor(raw / 2 ** (shift * 8)) & 0xff;
      }
    } else {
      const start = field.byte * 8 + field.bit;
      for (let offset = 0; offset < field.bits; offset += 1) {
        const position = start + offset;
        payload[Math.floor(position / 8)] |= (Math.floor(raw / 2 ** offset) % 2) << (position % 8);
      }
    }
  }
  return ["ok", payload];
}

export function encodeGenerated(
  messageKey: string,
  values: Readonly<Record<string, unknown>>,
  bus: string,
): EncodeResult {
  let message: MessageMetadata;
  try {
    message = generatedMessage(messageKey);
  } catch (error) {
    if (error instanceof RangeError && error.message.startsWith("unknown message")) return ["wrong_message_id"];
    throw error;
  }
  const instance = instanceFor(message, bus);
  if (instance === undefined) return ["wrong_message_id"];
  const [status, payload] = encodePayload(message, values);
  if (status !== "ok" || payload === undefined) return [status];
  return ["ok", frame(bus, instance.id, instance.frame_format, payload)];
}

export function decodeGenerated(messageKey: string, input: CanFrame): DecodeResult<Record<string, number>> {
  let message: MessageMetadata;
  try {
    message = generatedMessage(messageKey);
  } catch (error) {
    if (error instanceof RangeError && error.message.startsWith("unknown message")) return ["wrong_message_id"];
    throw error;
  }
  const instance = instanceFor(message, input.bus);
  if (instance === undefined) return ["wrong_message_id"];
  const status = validateFrame(input, {
    bus: instance.bus,
    id: instance.id,
    frameFormat: instance.frame_format,
    dlc: message.dlc,
  });
  if (status !== "ok") return [status];

  const values: Record<string, number> = {};
  for (const field of message.layout.fields ?? []) {
    let raw = 0;
    if (field.bit === 0 && field.bits % 8 === 0) {
      const width = field.bits / 8;
      for (let index = 0; index < width; index += 1) {
        const shift = message.byte_order === "big" ? width - index - 1 : index;
        raw += input.data[field.byte + index] * 2 ** (shift * 8);
      }
    } else {
      const start = field.byte * 8 + field.bit;
      for (let offset = 0; offset < field.bits; offset += 1) {
        const position = start + offset;
        raw += ((input.data[Math.floor(position / 8)] >> (position % 8)) & 1) * 2 ** offset;
      }
    }
    if (field.signed && raw >= 2 ** (field.bits - 1)) raw -= 2 ** field.bits;
    if (field.enum !== undefined && isExhaustiveEnum(field) && !(String(raw) in field.enum)) return ["invalid_enum"];
    const value = raw * (field.factor ?? 1) + (field.offset ?? 0);
    const [minimum, maximum] = limits(field);
    if (value < minimum || value > maximum) return ["value_out_of_range"];
    if (field.constant !== undefined && value !== field.constant) return ["constant_mismatch"];
    values[field.key] = value;
  }
  return ["ok", values];
}

export function decodeGeneratedInto(
  message: string,
  input: CanFrame,
  output: Record<string, unknown>,
): CodecStatus {
  const [status, value] = decodeGenerated(message, input);
  if (status === "ok" && value !== undefined) {
    for (const key of Object.keys(output)) delete output[key];
    Object.assign(output, value);
  }
  return status;
}

File: etrike_protocol/codecs/python/generated.py
from __future__ import annotations

import math
from collections.abc import Mapping, MutableMapping
from typing import Any

from protocol.generated.python.etrike_protocol import METADATA

from .types import CodecStatus, Frame, FrameFormat, validate_frame


def is_generated(message: str) -> bool:
    metadata = METADATA.get(message)
    return metadata is not None and metadata["codec"]["strategy"] == "generated"


def _message(message: str) -> dict[str, Any]:
    metadata = METADATA.get(message)
    if metadata is None:
        raise KeyError(message)
    if metadata["codec"]["strategy"] != "generated":
        raise ValueError(f"{message} selects a custom codec")
    return metadata


def _instance(metadata: Mapping[str, Any], bus: str) -> Mapping[str, Any] | None:
    matches = [item for item in metadata["instances"] if item["bus"] == bus]
    return matches[0] if len(matches) == 1 else None


def _limits(field: Mapping[str, Any]) -> tuple[float, float]:
    bits = field["bits"]
    if field.get("signed"):
        raw_minimum, raw_maximum = -(1 << (bits - 1)), (1 << (bits - 1)) - 1
    else:
        raw_minimum, raw_maximum = 0, (1 << bits) - 1
    factor, offset = field.get("factor", 1), field.get("offset", 0)
    return (
        field.get("min", raw_minimum * factor + offset),
        field.get("max", raw_maximum * factor + offset),
    )


def _encode_payload(metadata: Mapping[str, Any], values: Mapping[str, object]) -> tuple[CodecStatus, bytes | None]:
    payload = bytearray(metadata["dlc"])
    for field in metadata["layout"]["fields"]:
        value = field.get("constant", values.get(field["key"]))
        if not isinstance(value, (int, float)) or isinstance(value, bool) or not math.isfinite(value):
            return "value_out_of_range", None
        minimum, maximum = _limits(field)
        if value < minimum or value > maximum:
            return "value_out_of_range", None
        factor, offset = field.get("factor", 1), field.get("offset", 0)
        unrounded = (value - offset) / factor
        raw = round(unrounded)
        if not math.isclose(unrounded, raw, rel_tol=0, abs_tol=1e-9):
            return "value_out_of_range", None
        if "enum" in field and str(raw) not in field["enum"]:
            return "invalid_enum", None
        if field.get("signed") and raw < 0:
            raw += 1 << field["bits"]
        bits = field["bits"]
        if field["bit"] == 0 and bits % 8 == 0:
            width = bits // 8
            payload[field["byte"] : field["byte"] + width] = raw.to_bytes(width, metadata["byte_order"])
        else:
            start = field["byte"] * 8 + field["bit"]
            for offset_index in range(bits):
                position = start + offset_index
                payload[position // 8] |= ((raw >> offset_index) & 1) << (position % 8)
    return "ok", bytes(payload)


def encode(message: str, values: Mapping[str, object], *, bus: str) -> tuple[CodecStatus, Frame | None]:
    try:
        metadata = _message(message)
    except KeyError:
        return "wrong_message_id", None
    instance = _instance(metadata, bus)
    if instance is None:
        return "wrong_message_id", None
    status, payload = _encode_payload(metadata, values)
    if status != "ok":
        return status, None
    return "ok", Frame(bus, instance["id"], instance["frame_format"], payload)


def decode(message: str, frame: Frame) -> tuple[CodecStatus, dict[str, int | float] | None]:
    try:
        metadata = _message(message)
    except KeyError:
        return "wrong_message_id", None
    instance = _instance(metadata, frame.bus)
    if instance is None:
        return "wrong_message_id", None
    status = validate_frame(
        frame,
        bus=instance["bus"],
        can_id=instance["id"],
        frame_format=instance["frame_format"],
        dlc=metadata["dlc"],
    )
    if status != "ok":
        return status, None

    values: dict[str, int | float] = {}
    for field in metadata["layout"]["fields"]:
        bits = field["bits"]
        if field["bit"] == 0 and bits % 8 == 0:
            width = bits // 8
            raw = int.from_bytes(
                frame.data[field["byte"] : field["byte"] + width], metadata["byte_order"]
            )
        else:
            raw = 0
            start = field["byte"] * 8 + field["bit"]
            for offset_index in range(bits):
                position = start + offset_index
                raw |= ((frame.data[position // 8] >> (position % 8)) & 1) << offset_index
        if field.get("signed") and raw & (1 << (bits - 1)):
            raw -= 1 << bits
        if "enum" in field and str(raw) not in field["enum"]:
            return "invalid_enum", None
        value = raw * field.get("factor", 1) + field.get("offset", 0)
        minimum, maximum = _limits(field)
        if value < minimum or value > maximum:
            return "value_out_of_range", None
        if "constant" in field and value != field["constant"]:
            return "constant_mismatch", None
        values[field["key"]] = value
    return "ok", values


def decode_into(
    message: str, frame: Frame, output: MutableMapping[str, object]
) -> CodecStatus:
    status, value = decode(message, frame)
    if status == "ok" and value is not None:
        output.clear()
        output.update(value)
    return status


def decode_payload(
    message: str,
    payload: bytes,
    *,
    bus: str,
    frame_format: FrameFormat,
    can_id: int,
    dlc: int | None = None,
) -> tuple[CodecStatus, dict[str, int | float] | None]:
    return decode(message, Frame(bus, can_id, frame_format, payload, dlc))
